pkg/controller/master/master_controller.go

```diff
@@ -157,16 +182,114 @@ func (s *installStrategy) setupServiceAccountForJob(clusterNamespace string) (*k
 	return currentSA, nil
 }
 
+func (s *installStrategy) masterMachineSet(cluster *clustop.CombinedCluster) (*capi.MachineSet, error) {
+	masterMachineSetName := controller.MasterMachineSetName(cluster.Name)
+	return s.machineSetLister.MachineSets(cluster.Namespace).Get(masterMachineSetName)
+}
+
+func (s *installStrategy) availableMasterMachines(masterMachineSet *capi.MachineSet) ([]*capi.Machine, error) {
+	machines := []*capi.Machine{}
+	allMachines, err := s.machineLister.Machines(masterMachineSet.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	for _, machine := range allMachines {
+		if !metav1.IsControlledBy(machine, masterMachineSet) ||
+			machine.DeletionTimestamp != nil {
+			continue
+		}
+		hasPublicName, err := hasPublicDNS(machine)
+		if err != nil {
+			return nil, fmt.Errorf("error checking whether machine %s/%s has public DNS: %v", machine.Namespace, machine.Name, err)
+		}
+		if !hasPublicName {
```

Comment:
If one of the masters does not have a public name yet, should we wait for it before installing the control plane? It would seem to me to be akin to waiting for all the replicas to be created.

Reply:
We would wait, because the machine would not be included in the list of available machines and those would not match the replica count of the machineset.